[CUDA][ROCM ?] header incompatibility with gcc 12

Currently there is a header incompatiblity with gcc 12 : 
cf https://github.com/AdaptiveCpp/AdaptiveCpp/blob/develop/.github/workflows/linux.yml : 
```yaml
- name: uninstall gcc 12 development headers due to CUDA incompatibility
      if: matrix.os == 'ubuntu-22.04'
      run : |
        sudo apt remove libgcc-12-dev libstdc++-12-dev
```

using cuda 11.8, Debian clang version 16.0.6, gcc (Debian 12.2.0-14) 12.2.0
```
syclcc --hipsycl-targets='cuda:sm_70' --hipsycl-config-file=...
```
on the following file : 
```
#include <sycl/sycl.hpp>
int main(void){}
```

gives : 
```
acpp warning: No optimization flag was given, optimizations are disabled by default. Performance may be degraded. Compile with e.g. -O2/-O3 to enable optimizations.
In file included from sycl_include.cpp:1:
In file included from /scratch/tdavidcl/Shamrock_tests/cuda/OpenSYCL_comp/bin/../include/sycl/sycl.hpp:31:
In file included from /scratch/tdavidcl/Shamrock_tests/cuda/OpenSYCL_comp/include/sycl/../hipSYCL/sycl/sycl.hpp:50:
In file included from /scratch/tdavidcl/Shamrock_tests/cuda/OpenSYCL_comp/bin/../include/hipSYCL/glue/persistent_runtime.hpp:31:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/memory:77:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/shared_ptr.h:53:
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/shared_ptr_base.h:196:22: error: use of undeclared identifier 'noinline'; did you mean 'inline'?
      __attribute__((__noinline__))
                     ^
/usr/include/crt/host_defines.h:83:24: note: expanded from macro '__noinline__'
        __attribute__((noinline))
                       ^
In file included from sycl_include.cpp:1:
In file included from /scratch/tdavidcl/Shamrock_tests/cuda/OpenSYCL_comp/bin/../include/sycl/sycl.hpp:31:
In file included from /scratch/tdavidcl/Shamrock_tests/cuda/OpenSYCL_comp/include/sycl/../hipSYCL/sycl/sycl.hpp:50:
In file included from /scratch/tdavidcl/Shamrock_tests/cuda/OpenSYCL_comp/bin/../include/hipSYCL/glue/persistent_runtime.hpp:31:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/memory:77:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/shared_ptr.h:53:
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/shared_ptr_base.h:196:22: error: type name does not allow function specifier to be specified
/usr/include/crt/host_defines.h:83:24: note: expanded from macro '__noinline__'
        __attribute__((noinline))
                       ^
In file included from sycl_include.cpp:1:
In file included from /scratch/tdavidcl/Shamrock_tests/cuda/OpenSYCL_comp/bin/../include/sycl/sycl.hpp:31:
In file included from /scratch/tdavidcl/Shamrock_tests/cuda/OpenSYCL_comp/include/sycl/../hipSYCL/sycl/sycl.hpp:50:
In file included from /scratch/tdavidcl/Shamrock_tests/cuda/OpenSYCL_comp/bin/../include/hipSYCL/glue/persistent_runtime.hpp:31:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/memory:77:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/shared_ptr.h:53:
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/shared_ptr_base.h:196:22: error: expected expression
/usr/include/crt/host_defines.h:83:33: note: expanded from macro '__noinline__'
        __attribute__((noinline))
                                ^
3 errors generated when compiling for sm_70.
```

This issue is also referenced in : https://github.com/AdaptiveCpp/AdaptiveCpp/issues/939 , https://github.com/AdaptiveCpp/AdaptiveCpp/issues/852

-------

To quote an answer i got in the llvm discord

```
you can manually write over that file in your install directory, or potentially provide your 
own by the same name and make it be found first with -internal-isystem or something to your path

They should be in /install/lib/clang/<version>/include
```
maybe this can be implemented in acpp as a workaround for this issue, however it may depend on the include order, in that case 
quote : 
```
but you would need to do everything manually
I.e. -nobuiltininc -nogpuinc and then pass the resource directory includes manually 
via -isystem /yours/ -isytem /install//resource/dir/include or something
I recommend using -### to check the order these are showing up in 
```

--- a/.github/workflows/linux.yml
+++ b/.github/workflows/linux.yml
@@ -39,10 +39,6 @@ jobs:
     - name: Install TBB for PSTL tests
       run : |
         sudo apt install libtbb-dev
-    - name: uninstall gcc 12 development headers due to CUDA incompatibility
-      if: matrix.os == 'ubuntu-22.04'
-      run : |
-        sudo apt remove libgcc-12-dev libstdc++-12-dev
     - name: install Level Zero
       run : |
         wget https://github.com/oneapi-src/level-zero/releases/download/v1.2.3/level-zero-devel_1.2.3+u18.04_amd64.deb

--- a/include/hipSYCL/std/hiplike/bits/shared_ptr_base.h
+++ b/include/hipSYCL/std/hiplike/bits/shared_ptr_base.h
@@ -1,0 +1,11 @@
+// Taken from https://reviews.llvm.org/D149364
+
+// CUDA headers define __noinline__ which interferes with libstdc++'s use of
+// `__attribute((__noinline__))`. In order to avoid compilation error,
+// temporarily unset __noinline__ when we include affected libstdc++ header.
+
+#pragma push_macro("__noinline__")
+#undef __noinline__
+#include_next "bits/shared_ptr_base.h"
+
+#pragma pop_macro("__noinline__")

